### ISSUE TITLE
Incorrect outcome probabiliy balancing.

### ISSUE THREAD

**Issue by bh2smith** [protofire/gnosis-conditional-exchange#597]
When trying to put money on the last outcome here:

![image](https://user-images.githubusercontent.com/11778116/79948138-92f72f80-8473-11ea-9fa5-7bf0c81b87b2.png)

the first two outcomes probability is also increasing. Realistically, what should happen is that if you are bidding one price up, everything else MUST go down

![image](https://user-images.githubusercontent.com/11778116/79948226-bb7f2980-8473-11ea-96bd-bc283c94193a.png)

**Comment by fvictorio** [comment]
I can confirm that this happens, but if you go ahead and make a purchase, then the percentages do change in that way, so the prediction is correct.

I think this must be just something that happens in these edge cases, where the odds are very skewed to one outcome. @cag can you confirm/deny this?

**Comment by bh2smith** [comment]
Yes, there was a similar UI bug in the interface of the Helena prediction markets. This happened exactly in the outcome estimations when trying to buy an outcome that was already nearly maxed out (at ~97%).

**Comment by cag** [comment]
That is really curious... Off the cuff, I wouldn't expect rounding stuff like this to happen until ~99% mark because of USDC having 6 decimal points and the price at somewhere around .01 ish.

If it's a rounding error, it might have to do with the way multiplies are not grouped before divisions in the cost calculations, but rather alternated to avoid overflow due to the lack of floating point support. I think we didn't opt for a soft float impl because it's not been battle-tested, and because it would have been a large audit burden. It would be interesting to see if other such implementations have such an issue.

Anyway, my best guess at this point is that this is something related to how intermediate values get rounded, but this will require further investigation to confirm.

**Comment by fvictorio** [comment]
Not sure I'm following this. The computation in the UI is done with arbitrary precision, so there shouldn't be any errors. You mean it's the change in holdings in the market maker that is the problem?

**Comment by cag** [comment]
That was my initial hypothesis, but now I'm not so sure.

**Comment by cag** [comment]
Okay, the reason is deeper than rounding errors. Basically, the rounding errors are fine, and everything is implemented to spec. This behavior will persist even if we're using real numbers.

I think it has to do with the marginal price expressions used here *not* being partial derivatives of a scalar field, or equivalently, they form a non-conservative vector field. Maybe there are more "correct" ways of calculating probability/marginal price... For reference referring to the notes I took, the price expression would be something like dx/dy (ratio of a small amount of collateral used to buy to a corresponding amount of outcome token bought).

**Comment by fvictorio** [comment]
So there's nothing we can do here, is it?

The only thing I can think of is that, if the new price is bigger for an outcome that is not being bought (or for an outcome that is being sold), or lower for an outcome that is not being sold (or an outcome that is being bought), then we don't show the change in the UI. But that might be a lot of work for a kind of edge case.

**Comment by cag** [comment]
I had some additional notes on this, but basically, there's [a paper on constant function market makers](https://web.stanford.edu/~guillean/papers/constant_function_amms.pdf) where it is mentioned that the marginal prices for an N-dimensional market should be proportional to the gradient of the constant function (or the product in this case).

The marginal prices/probabilities in this case are indeed proportional to the gradient, so yeah, the reason is not that the incorrect expression is used here, but rather something else...
